### DISCUSSION TITLE
Toggle option for powering monitor on and off.

### DISCUSSION THREAD

**Discussion by deadbody13** [YaLTeR/niri#724] [Ideas]
When I play games, watch a movie, or just want a single monitor to focus on something I've been going into the config and add `off` to one of the monitor output sections to make it so only one screen is on. It reduces ambient light and helps me focus.

It'd be cool to have a command that I can bind to toggle a monitor between on and off so I didn't have to go into the config to do so. Something like:
```
binds {
    Super+p { spawn "niri" "msg" "toggle-monitor" "DP-1" }
}
```
Or maybe a way to pipe this into fuzzel so that we have a Windows-esque Primary Only, Secondary Only, Extend, and Duplicate option.

**Comment by YaLTeR** [comment]
You could make a script that checks the current state in `niri msg --json outputs` and calls `niri msg output DP-1 on` or `off` accordingly, then spawn that script from the bind.

> Duplicate

Output mirroring isn't implemented in niri itself yet and it is quite complex to implement too, so probably not in the near future. You can use wl-mirror though.

**Comment by bphenriques** [comment]
@deadbody13 I have been working on this and I can share what I have (atm):
```sh
# Functions that depend on the window manager
exists_display() { niri msg --json outputs | jq --exit-status --arg DISPLAY "$1" '.[$DISPLAY]' > /dev/null; }
is_display_on() { niri msg --json outputs | jq --exit-status --arg DISPLAY "$1" '.[$DISPLAY] | .current_mode' > /dev/null; }
toggle() { exists_display "$1" && niri msg output "$1" "$2"; }

# Utility functions
profile_internal_name() { echo -n "Internal display"; }
profile_internal_valid() { exists_display eDP-1; }
profile_internal_set() { toggle HDMI-A-1 off && toggle eDP-1 on; }

profile_external_name() { echo -n "External display"; }
profile_external_valid() { exists_display HDMI-A-1; }
profile_external_set() { toggle eDP-1 off && toggle HDMI-A-1 on; }

profile_extend_name() { echo -n "Extend displays"; }
profile_extend_valid() { exists_display "eDP-1" && exists_display "HDMI-A-1"; }
profile_extend_set() { toggle eDP-1 on && toggle HDMI-A-1 on; }

DEFAULT="$(profile_internal_name)"

# shellcheck disable=SC2005
options() {
  profile_internal_valid && echo "$(profile_internal_name)"
  profile_external_valid && echo "$(profile_external_name)"
  profile_extend_valid && echo "$(profile_extend_name)"
}

dmenu() {
  options | fuzzel --dmenu --placeholder "Current: ${DEFAULT}" --lines 3
}

# FIXME
notify() {
  notify-send "Display layout" "Set to ..."
}

case "$(dmenu)" in
  "$(profile_internal_name)") profile_internal_set && notify ;;
  "$(profile_external_name)") profile_external_set && notify ;;
  "$(profile_extend_name)")   profile_extend_set && notify ;;
esac
```

The only thing missing from me is having a way to (separately) detect when the HDMI cable is disconnected. The `ipc` stream does not show that as far as I can tell and I can open a separate request.

Edit: I went a different route.

**Comment by solarfl4re** [comment]
I usually keep my laptop monitor off when my external display is connected, but sometimes want to toggle it on; I wrote a small script I can spawn with a key bind:
```bash
# is eDP-1 disabled?
if [[ $(niri msg --json outputs | jq '."eDP-1".current_mode == null') == "true" ]]; then
  # then enable it
  niri msg output eDP-1 on
else
  # or if it's on, turn it off
  niri msg output eDP-1 off
fi
```

**Comment by deadbody13** [comment]
Atm I have a shell script that launches fuzzel with 3 basic output options. This is on desktop so I haven't had to worry about clam mode or anything yet.
```
#!/bin/bash

input=$(printf "suspend\0icon\x1fsystem-suspend\nshutdown\0icon\x1fsystem-shutdown\nreboot\0icon\x1fsystem-restart" | fuzzel --dmenu --icon-theme=Obsidian-Silver-Light)

case "$input" in
	suspend)
		systemctl suspend
		;;
	shutdown)
		shutdown now
		;;
	reboot)
		reboot
		;;
esac
```

Edit: I'm an idiot, this is my power script.... 
THIS is my screen control script:
```
#!/bin/bash
### select screen power

input=$(printf "primary only\nsecondary only\nextend" | fuzzel --dmenu)

output_1="DP-3"
output_2="DP-4"

case "$input" in
	"primary only")
		niri msg output $output_1 off ; niri msg output $output_2 on
		;;
	"secondary only")
		niri msg output $output_1 on ; niri msg output $output_2 off
		;;
	extend)
		niri msg output $output_1 on ; niri msg output $output_2 on
		;;
esac
```

**Comment by hyperfocusedblur** [comment]
> When I play games, watch a movie, or just want a single monitor to focus on something I've been going into the config and add `off` to one of the monitor output sections to make it so only one screen is on. It reduces ambient light and helps me focus.
> 
> It'd be cool to have a command that I can bind to toggle a monitor between on and off so I didn't have to go into the config to do so. Something like:
> 
> ```
> binds {
>     Super+p { spawn "niri" "msg" "toggle-monitor" "DP-1" }
> }
> ```
> 
> Or maybe a way to pipe this into fuzzel so that we have a Windows-esque Primary Only, Secondary Only, Extend, and Duplicate 

you can put this oneliner as a bind in niri/config to toggle any particular display:
`niri msg outputs | grep <name_of_display> -n2 | grep Disabled -q && niri msg output <name_of_display> on || niri msg output <name_of_display> off`

for example: 	
`XF86Display allow-when-locked=false { spawn-sh "niri msg outputs | grep eDP-1 -n2 | grep Disabled -q && niri msg output eDP-1 on || niri msg output eDP-1 off"; }`
